sources/sql/src/lib.rs

```diff
@@ -106,64 +118,377 @@ fn rewrite_table_scans(plan: &LogicalPlan) -> Result<LogicalPlan> {
-            let subquery_alias = LogicalPlan::SubqueryAlias(SubqueryAlias::try_new(
-                Arc::new(LogicalPlan::TableScan(new_table_scan)),
-                original_table_name,
-            )?);
-
-            return Ok(subquery_alias);
+            return Ok(LogicalPlan::TableScan(new_table_scan));
         } else {
             return Ok(plan.clone());
         }
     }
 
-    let mut new_expressions = vec![];
-    for expression in plan.expressions() {
-        new_expressions.push(rewrite_table_scans_in_subqueries(expression)?);
-    }
-
     let rewritten_inputs = plan
         .inputs()
         .into_iter()
-        .map(rewrite_table_scans)
+        .map(|i| rewrite_table_scans(i, known_rewrites))
         .collect::<Result<Vec<_>>>()?;
 
+    let mut new_expressions = vec![];
+    for expression in plan.expressions() {
+        let new_expr = rewrite_table_scans_in_expr(expression.clone(), known_rewrites)?;
+        new_expressions.push(new_expr);
+    }
+
     let new_plan = plan.with_new_exprs(new_expressions, rewritten_inputs)?;
 
     Ok(new_plan)
 }
 
-fn rewrite_table_scans_in_subqueries(expr: Expr) -> Result<Expr> {
+fn rewrite_table_scans_in_expr(
+    expr: Expr,
+    known_rewrites: &mut HashMap<TableReference, TableReference>,
+) -> Result<Expr> {
     match expr {
         Expr::ScalarSubquery(subquery) => {
-            let new_subquery = rewrite_table_scans(&subquery.subquery)?;
+            let new_subquery = rewrite_table_scans(&subquery.subquery, known_rewrites)?;
+            let outer_ref_columns = subquery
+                .outer_ref_columns
+                .into_iter()
+                .map(|e| rewrite_table_scans_in_expr(e, known_rewrites))
+                .collect::<Result<Vec<Expr>>>()?;
             Ok(Expr::ScalarSubquery(Subquery {
                 subquery: Arc::new(new_subquery),
-                outer_ref_columns: subquery.outer_ref_columns,
+                outer_ref_columns,
             }))
         }
         Expr::BinaryExpr(binary_expr) => {
-            let left = rewrite_table_scans_in_subqueries(*binary_expr.left)?;
-            let right = rewrite_table_scans_in_subqueries(*binary_expr.right)?;
+            let left = rewrite_table_scans_in_expr(*binary_expr.left, known_rewrites)?;
+            let right = rewrite_table_scans_in_expr(*binary_expr.right, known_rewrites)?;
             Ok(Expr::BinaryExpr(BinaryExpr::new(
                 Box::new(left),
                 binary_expr.op,
                 Box::new(right),
             )))
         }
-        _ => {
-            tracing::debug!("rewrite_table_scans_in_subqueries: no match for expr={expr:?}",);
-            Ok(expr)
+        Expr::Column(col) => {
+            if let Some(rewrite) = col.relation.as_ref().and_then(|r| known_rewrites.get(r)) {
+                Ok(Expr::Column(Column::new(Some(rewrite.clone()), &col.name)))
```

Comment:
Column references that reference a column we rewrote earlier should get the new reference.